src/Features/Core/Portable/AddRequiredParentheses/AbstractAddRequiredParenthesesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,98 @@
+        {
+        }
+
+        public sealed override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public sealed override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected sealed override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeSyntax, GetSyntaxNodeKinds());
+
+        protected abstract ImmutableArray<TLanguageKindEnum> GetSyntaxNodeKinds();
+        protected abstract ISyntaxFactsService GetSyntaxFactsService();
+        protected abstract int GetPrecedence(TBinaryExpressionSyntax binaryExpression);
+        protected abstract PrecedenceKind GetPrecedenceKind(TBinaryExpressionSyntax binaryExpression, SemanticModel semanticModel);
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+        {
+            var syntaxTree = context.SemanticModel.SyntaxTree;
+            var cancellationTokan = context.CancellationToken;
+            var optionSet = context.Options.GetDocumentOptionSetAsync(syntaxTree, cancellationTokan).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var binaryExpression = (TBinaryExpressionSyntax)context.Node;
+            if (!(binaryExpression.Parent is TBinaryExpressionSyntax parentBinary))
+            {
+                return;
+            }
+
+            if (GetPrecedence(binaryExpression) == GetPrecedence(parentBinary))
+            {
+                return;
+            }
+
+            var precedenceKind = GetPrecedenceKind(parentBinary, context.SemanticModel);
+
+            var preference = optionSet.GetOption(GetLanguageOption(precedenceKind), binaryExpression.Language);
+            if (preference.Value != ParenthesesPreference.RequireForPrecedenceClarity)
+            {
+                return;
+            }
+
+            var additionalLocations = ImmutableArray.Create(binaryExpression.GetLocation());
+            var precedence = GetPrecedence(binaryExpression);
+
+            // In a case like "a + b * c * d", we'll add parens to make "a + (b * c * d)".
+            // To make this user experience more pleasant, we will place the diagnostic on
+            // both *'s.
+            AddDiagnostics(
+                context, binaryExpression, precedence,
+                preference.Notification.Value, additionalLocations);
+        }
+
+        private void AddDiagnostics(
+            SyntaxNodeAnalysisContext context, TBinaryExpressionSyntax binaryExpressionOpt,
+            int precedence, DiagnosticSeverity severity, ImmutableArray<Location> additionalLocations)
+        {
+            if (binaryExpressionOpt != null && GetPrecedence(binaryExpressionOpt) == precedence)
+            {
+                var syntaxFacts = GetSyntaxFactsService();
```

Comment:
why temporary?

Reply:
general coding pattern that we follow elsewhere.  We usually place services into a local, even if we only use them once.

Reply:
It also helps with readability to me.  ```var x = syntaxFacts.XXX(``` is clearer to me than ```var x = GetSyntaxFactsService().XXX(```.

Reply:
ok